File: bst_closest_value/closest.py
def findClosestValueInBst(tree, target):
    return findClosest(tree, target, tree.value)


def findClosest(tree, target, closest):
    if not tree:
        return closest
    if tree.value == target:
        return tree.value
    if abs(tree.value - target) < abs(closest - target):
        closest = tree.value
    if tree.value < target:
        return findClosest(tree.right, target, closest)
    else:
        return findClosest(tree.left, target, closest)


class BST:
    def __init__(self, value):
        self.value = value
        self.left = None
        self.right = None


def test_case_1():
    root = BST(10)
    root.left = BST(5)
    root.left.left = BST(2)
    root.left.left.left = BST(1)
    root.left.right = BST(5)
    root.right = BST(15)
    root.right.left = BST(13)
    root.right.left.right = BST(14)
    root.right.right = BST(22)
    expected = 22
    actual = findClosestValueInBst(root, 20)
    print("{} vs {}".format(expected, actual))


if __name__ == '__main__':
    test_case_1()

File: bst_closest_value/closest.go
package main

import (
	"fmt"
	"io"
	"math"
	"os"
)

type Node struct {
	Value int
	Left  *Node
	Right *Node
}

type BST struct {
	Root *Node
}

func (node *Node) insert(value int) {
	if node == nil {
		return
	} else if value < node.Value {
		if node.Left == nil {
			node.Left = &Node{Value: value, Left: nil, Right: nil}
		} else {
			node.Left.insert(value)
		}
	} else {
		if node.Right == nil {
			node.Right = &Node{Value: value, Left: nil, Right: nil}
		} else {
			node.Right.insert(value)
		}
	}
}

func (tree *BST) insert(value int) *BST {
	if tree.Root == nil {
		tree.Root = &Node{Value: value, Left: nil, Right: nil}
	} else {
		tree.Root.insert(value)
	}
	return tree
}

func print(w io.Writer, node *Node, ns int, ch rune) {
	if node == nil {
		return
	}
	for i := 0; i < ns; i++ {
		fmt.Fprint(w, " ")
	}
	fmt.Fprintf(w, "%c:%v\n", ch, node.Value)
	print(w, node.Left, ns+2, 'L')
	print(w, node.Right, ns+2, 'R')
}

func findClosest(node *Node, value int, closest int) int {
	if node == nil {
		return closest
	}
	if node.Value == value {
		return node.Value
	}
	if math.Abs(float64(node.Value-value)) < math.Abs(float64(closest-value)) {
		closest = node.Value
	}
	if node.Value < value {
		return findClosest(node.Right, value, closest)
	} else {
		return findClosest(node.Left, value, closest)
	}
}

func (tree *BST) findClosestBSTValue(value int) int {
	return findClosest(tree.Root, value, tree.Root.Value)
}

func main() {
	tree := &BST{}
	tree.insert(100).
		insert(-20).
		insert(-50).
		insert(-15).
		insert(-60).
		insert(50).
		insert(60).
		insert(55).
		insert(85).
		insert(15).
		insert(5).
		insert(-10)
	print(os.Stdout, tree.Root, 0, 'M')
	to_find := 35
	fmt.Printf("Closest to %d is %d\n", to_find, tree.findClosestBSTValue(to_find))
}
